Mathlib/Order/UpperLower/Basic.lean

```diff
@@ -946,6 +983,24 @@ theorem compl_iInf₂ (f : ∀ i, κ i → LowerSet α) :
     (⨅ (i) (j), f i j).compl = ⨅ (i) (j), (f i j).compl := by simp_rw [LowerSet.compl_iInf]
 #align lower_set.compl_infi₂ LowerSet.compl_iInf₂
 
+-- TODO: Introduce an `IsMaxOn` predicate?
+/-- Remove a maximal element from a lower set. -/
+def erase (s : LowerSet α) (a : α) (has : ∀ b ∈ s, a ≤ b → b = a) : LowerSet α where
```

Comment:
Do you think it's sensible to define this as `s` if `has` fails, so the definition doesn't require a proof obligation, ie is less partial?

Reply:
I think this will be strictly less practical, and my reasoning is that:
* this proof obligation is a non-obvious condition
* with the current design, it gets carried around instead of needing to be resupplied every time

However, we could also define `s.erase a` as "the biggest sublower set of `s` which does not contain `a`", namely `s \ UpperSet.Ici a`, or even `s.sdiff t` as `s \ UpperClosure t`. The covering properties would not hold in general, but that might still be enough for my applications. What do you think?

Reply:
Yeah, this sounds like a better option; isn't that then just `UpperClosure (s  \ {a})`?

Reply:
You mean `compl (upperClosure (insert a (compl s)))`? If so, yes.

Reply:
Oops I thought this was making an upper set for some reason! I meant something in terms of LowerClosure instead, but probably equal to what you wrote

Reply:
I don't think you can get that using `lowerClosure` anywhere, since you want to remove all the elements of `s` that are greater than `a`.

Reply:
Done. What do you think?